substrate/frame/staking/src/lib.rs

```diff
@@ -1328,18 +1350,117 @@ impl<T: Config, const DISABLING_LIMIT_FACTOR: usize> DisablingStrategy<T>
 				Pallet::<T>::current_era().unwrap_or_default(),
 				slash_era
 			);
-			return None
+			return DisablingDecision { disable: None, reenable: None }
 		}
 
 		let offender_idx = if let Some(idx) = active_set.iter().position(|i| i == offender_stash) {
 			idx as u32
 		} else {
 			log!(debug, "Won't disable: offender not in active set",);
-			return None
+			return DisablingDecision { disable: None, reenable: None }
 		};
 
 		log!(debug, "Will disable {:?}", offender_idx);
 
-		Some(offender_idx)
+		DisablingDecision { disable: Some(offender_idx), reenable: None }
+	}
+}
+
+/// Implementation of [`DisablingStrategy`] which disables validators from the active set up to a
+/// limit and if the limit is reached and the new offender is higher (bigger punishment/severity)
+/// then it re-enables to lowest offender to free up space for the new offender.
+///
+/// An extension of [`UpToLimitDisablingStrategy`].
+pub struct UpToLimitWithReEnablingDisablingStrategy<const DISABLING_LIMIT_FACTOR: usize = 3>;
+
+impl<const DISABLING_LIMIT_FACTOR: usize>
+	UpToLimitWithReEnablingDisablingStrategy<DISABLING_LIMIT_FACTOR>
+{
+	/// Disabling limit calculated from the total number of validators in the active set. When
+	/// reached re-enabling logic might kick in.
+	pub fn disable_limit(validators_len: usize) -> usize {
+		validators_len
+			.saturating_sub(1)
+			.checked_div(DISABLING_LIMIT_FACTOR)
+			.unwrap_or_else(|| {
+				defensive!("DISABLING_LIMIT_FACTOR should not be 0");
+				0
+			})
```

Comment:
nit: Since this logic is shared between `UpToLimitWithReEnablingDisablingStrategy` and `UpToLimitDisablingStrategy` it's worth to extract it in a separate function.